crates/job-queue-libsql/src/lib.rs

```diff
@@ -0,0 +1,414 @@
+                        .execute(sql_query, ())
+                        .await
+                        .with_context(|| format!("failed to execute migration {}", migration_id))?;
+
+                    trace!(migration_id, "Applied migration");
+
+                    self.update_user_version(migration_id + 1)
+                        .await
+                        .context("failed to update user version")?;
+                }
+
+                info!("Migration complete");
+            }
+            None => {
+                warn!(user_version, "user_version is set to an unexpected value");
+            }
+            _ => {
+                debug!(user_version, "Database is already up to date");
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn query_user_version(&self) -> anyhow::Result<usize> {
+        let sql_query = "PRAGMA user_version";
+
+        trace!(%sql_query, "Query user_version");
+
+        let row = self
+            .conn
+            .query(sql_query, ())
+            .await
+            .context("failed to execute SQL query")?
+            .next()
+            .await
+            .context("failed to read the row")?
+            .context("no row returned")?;
+
+        let value = row.get::<u64>(0).context("failed to read user_version value")?;
+
+        Ok(usize::try_from(value).expect("number not too big"))
+    }
+
+    async fn update_user_version(&self, value: usize) -> anyhow::Result<()> {
+        let value = u64::try_from(value).expect("number not too big");
+
+        let sql_query = format!("PRAGMA user_version = {value}");
+
+        trace!(%sql_query, "Update user_version");
+
+        self.conn
+            .execute(&sql_query, ())
+            .await
+            .context("failed to execute SQL query")?;
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl JobQueue for LibSqlJobQueue {
+    async fn setup(&self) -> anyhow::Result<()> {
+        self.apply_pragmas().await?;
+        self.migrate().await?;
+        Ok(())
+    }
+
+    async fn reset_claimed_jobs(&self) -> anyhow::Result<()> {
+        let sql_query = "UPDATE job_queue SET status = :queued_status WHERE status = :running_status";
+
+        let params = (
+            (":running_status", JobStatus::Running as u32),
+            (":queued_status", JobStatus::Queued as u32),
+        );
+
+        trace!(%sql_query, ?params, "Reset claimed jobs");
+
+        let changed_count = self
+            .conn
+            .execute(sql_query, params)
+            .await
+            .context("failed to execute SQL query")?;
+
+        trace!(changed_count, "Jobs reset with success");
+
+        Ok(())
+    }
+
+    async fn push_job(&self, job: &DynJob, schedule_for: Option<OffsetDateTime>) -> anyhow::Result<()> {
+        let sql_query = "INSERT INTO job_queue
+            (id, scheduled_for, failed_attempts, status, name, def)
+            VALUES (:id, :scheduled_for, :failed_attempts, :status, :name, jsonb(:def))";
```

Comment:
Cool, SQLite finally added a dedicated binary JSON type! 👍

Reply:
Yeah, seems convenient, it’s possible to perform query and indexing on it even!